pkg/aggregator/ckey/key.go

```diff
@@ -27,49 +23,88 @@ import (
-// but the Go compiler is not inlining xxhash sum methods whereas it is inlining
-// the murmur3 implementation, providing better performances overall.
 // Note that benchmarks against fnv1a did not provide better performances (no inlining).
 type ContextKey uint64
 
+/*
+#
+#// KeyGenerator generates key
+#// Not safe for concurrent usage
+#type KeyGenerator struct {
+#	buf []byte
+#}
+#
+#// NewKeyGenerator creates a new key generator
+#func NewKeyGenerator() *KeyGenerator {
+#	return &KeyGenerator{
+#		buf: make([]byte, 0, 1024),
+#	}
+#}
+#
+#// Generate returns the ContextKey hash for the given parameters.
+#// The tags array is sorted in place to avoid heap allocations.
+#func (g *KeyGenerator) Generate(name, hostname string, tags []string) ContextKey {
+#	g.buf = g.buf[:0]
+#
+#	// Sort the tags in place. For typical tag slices, we use
+#	// the in-place insertion sort to avoid heap allocations.
+#	// We default to stdlib's sort package for longer slices.
+#	// See `pkg/util/sort.go` for info on the threshold.
+#	if len(tags) < util.InsertionSortThreshold {
+#		util.InsertionSort(tags)
+#	} else {
+#		sort.Strings(tags)
+#	}
+#
+#	g.buf = append(g.buf, name...)
+#	g.buf = append(g.buf, ',')
+#	for i := 0; i < len(tags); i++ {
+#		g.buf = append(g.buf, tags[i]...)
+#		g.buf = append(g.buf, ',')
+#	}
+#	g.buf = append(g.buf, hostname...)
+#
+#	return ContextKey(murmur3.Sum64(g.buf))
+#}
+#
+*/
+
+// NewKeyGenerator creates a new key generator
+func NewKeyGenerator() *KeyGenerator {
+	return &KeyGenerator{}
+}
+
 // KeyGenerator generates key
 // Not safe for concurrent usage
 type KeyGenerator struct {
-	buf []byte
-}
+	intb uint64
 
-// NewKeyGenerator creates a new key generator
-func NewKeyGenerator() *KeyGenerator {
-	return &KeyGenerator{
-		buf: make([]byte, 0, 1024),
-	}
+	sli [512]uint64
+	idx int
 }
 
 // Generate returns the ContextKey hash for the given parameters.
 // The tags array is sorted in place to avoid heap allocations.
 func (g *KeyGenerator) Generate(name, hostname string, tags []string) ContextKey {
-	g.buf = g.buf[:0]
+	g.intb = 0xc6a4a7935bd1e995
+	g.intb = g.intb ^ murmur3.StringSum64(name)
+	g.intb = g.intb ^ murmur3.StringSum64(hostname)
 
-	// Sort the tags in place. For typical tag slices, we use
-	// the in-place insertion sort to avoid heap allocations.
-	// We default to stdlib's sort package for longer slices.
-	// See `pkg/util/sort.go` for info on the threshold.
-	if len(tags) < util.InsertionSortThreshold {
-		util.InsertionSort(tags)
-	} else {
-		sort.Strings(tags)
-	}
+	g.idx = 0
 
-	g.buf = append(g.buf, name...)
-	g.buf = append(g.buf, ',')
-	for i := 0; i < len(tags); i++ {
-		g.buf = append(g.buf, tags[i]...)
-		g.buf = append(g.buf, ',')
+OUTER:
+	for i := range tags {
+		h := murmur3.StringSum64(tags[i])
+		for j := 0; j < g.idx; j++ {
+			if g.sli[j] == h {
+				continue OUTER // we do not want to xor multiple times the same tag
+			}
```

Comment:
Maybe some sort of hashset would perform better here? Something like this:
```suggestion
		j := h & 511
		for {
			if g.seen[j] == nil {
				g.seen[j] = &tags[i]
				break
			} else if *g.seen[j] == tags[i] {
				continue OUTER
			} else {
				j = (j + 1) & 511
			}
		}
```

Reply:
That is interesting, especially since it is on a very small set of tags (< 200), the risk of collision may not be that high. I'll give it a shot in benchmarks!